samples/snippets/src/main/java/VerifyNonGoogleIdToken.java

```diff
@@ -0,0 +1,100 @@
+
+public class VerifyNonGoogleIdToken {
+
+  public static void main(String[] args)
+      throws IOException, ExecutionException, InterruptedException, GeneralSecurityException, JwkException {
+    // TODO(Developer): Replace the below variables before running the code.
+    // The non-Google ID token to verify.
+    String idToken = "id-token";
+
+    // The service name for which the id token is requested. Service name refers to the
+    // logical identifier of an API service, such as "pubsub.googleapis.com".
+    String targetAudience = "pubsub.googleapis.com";
+
+    // To verify non-google tokens, get the Json Web Key endpoint (jwk).
+    // OpenID Connect allows the use of a "Discovery document," a JSON document found at a
+    // well-known location containing key-value pairs which provide details about the
+    // OpenID Connect provider's configuration.
+    // For more information on validating the jwt, see: https://developers.google.com/identity/protocols/oauth2/openid-connect#validatinganidtoken
+    //
+    // Here, we validate Google's token using Google's OpenID Connect service (jwkUrl).
+    // For more information on jwk,see: https://auth0.com/docs/secure/tokens/json-web-tokens/json-web-key-sets
+    String jwkUrl = "https://www.googleapis.com/oauth2/v3/certs";
+
+    verifyNonGoogleIdToken(idToken, targetAudience, jwkUrl);
+  }
+
+  // Verify a non-google id token. Here, we are using the Google's jwk. The procedure is the same
+  // even if the jwk is from a different provider.
+  public static void verifyNonGoogleIdToken(String idToken, String targetAudience,
+      String jwkUrl)
+      throws IOException, JwkException {
+
+    // Start verification.
+    DecodedJWT jwt = JWT.decode(idToken);
+
+    // Check if the token has expired.
+    if (jwt.getExpiresAt().before(Calendar.getInstance().getTime())) {
+      System.out.println("Token already expired..");
+      return;
+    }
+
+    // Construct the jwkProvider from the provided jwkURL.
+    JwkProvider jwkProvider = new UrlJwkProvider(new URL(jwkUrl));
+    // Get the jwk from the provided key id.
+    Jwk jwk = jwkProvider.get(jwt.getKeyId());
+    // Retrieve the public key and use that to create an instance of the Algorithm.
+    Algorithm algorithm = Algorithm.RSA256((RSAPublicKey) jwk.getPublicKey(), null);
+    // Create the verifier with the algorithm and target audience.
+    JWTVerifier jwtVerifier = JWT.require(algorithm).withAudience(targetAudience).build();
+
+    boolean isVerified = true;
+    try {
+      // Verify the obtained id token. This is done at the receiving end of the OIDC endpoint.
+      jwt = jwtVerifier.verify(idToken);
+    } catch (JWTVerificationException e) {
+      System.out.println("Could not verify Signature: " + e.getMessage());
+      isVerified = false;
+    }
+
+    if (isVerified) {
+      System.out.println("Id token verified.");
+      return;
+    }
+    System.out.println("Unable to verify ID token.");
+  }
```

Comment:
Our TokenVerifier class supports loading a [public key explicitly](https://github.com/googleapis/google-auth-library-java/blob/ef17ed8bf2c91e7bd66d26f039f029de6af3814a/oauth2_http/java/com/google/auth/oauth2/TokenVerifier.java#L238) or [via a certificates URL](https://github.com/googleapis/google-auth-library-java/blob/ef17ed8bf2c91e7bd66d26f039f029de6af3814a/oauth2_http/java/com/google/auth/oauth2/TokenVerifier.java#L215).

We just default to loading from Google's public certificates endpoint.

Reply:
Thanks Jeff. Would you recommend we also demonstrate using that?

Reply:
I would recommend that we use our first-party implementation and not require extra third-party dependencies.